### ISSUE TITLE
Using create_context with LWS_SERVER_OPTION_EXPLICIT_VHOSTS

### ISSUE THREAD

**Issue by mfallavol** [warmcat/libwebsockets#2604]
I am trying to use libwebsockets as a protocol translator between a custom protocol on a raw TCP port and a websocket server.  I have the websocket client aspect working using the following initialization:

	memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
	info.port = CONTEXT_PORT_NO_LISTEN;
	info.protocols = protocols;
	context = lws_create_context(&info);

I may be mistaken but I believe in order to add the raw server port I have to add a new vhost with the new port.  Therefore as a first step I thought I'd just get the current code working with an explicit vhost.  I've changed the above initialization to this:

	memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
	info.options = LWS_SERVER_OPTION_EXPLICIT_VHOSTS;
	context = lws_create_context(&info);
	if (!context)
	{
		lwsl_err("lws init failed\n");
		return 1;
	}

	// Websocket vhost (client)
	memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
	info.port = CONTEXT_PORT_NO_LISTEN;
	info.protocols = protocols;
	pVhost = lws_create_vhost(context, &info);		

Though it runs, there seems to be a lot of new logging for what I believe is a secure streams vhost that I did not intend to create.  There is a new tag on startup:

`N: __lws_lc_tag:  ++ [vh|1|_ss_default||-1] (2)`

and a lot more logging about LWSSSCS_CREATING -> LWSSCS_POLL and captive_portal_detect.

I don't need the secure streams vhost.  Is there a way to get rid of it?  Am I taking the correct approach for adding the raw  socket?  I was intending to add the new (raw) vhost with this:

	// raw server on port 5310
	memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
	info.port = 5310;
	info.protocols = protocols;	
	info.options = LWS_SERVER_OPTION_ONLY_RAW;
	lws_create_vhost(context, &info);

I'm a little confused as to whether or not I should have a single protocol structure and if info.protocols should point to the same structure for both vhosts.  Since there is a single event loop I am assuming that they should be the same.

**Comment by lws-team** [comment]
> I may be mistaken but I believe in order to add the raw server port I have to add a new vhost with the new port. 

Yes basically the listen socket exists in the vhost object.  So if you want n listeners, you want n vhosts.

> Therefore as a first step I thought I'd just get the current code working with an explicit vhost.

There are some differences depending on whether you just let lws do its backwards-compatible thing or not.  SS server depends on EXPLICIT_VHOSTS, you're giving that now so it can start up the serving stuff.  If you don't want to use SS at all, you can disable it at cmake with `-DLWS_WITH_SECURE_STREAMS=0`, it's on by default nowadays.

Aside from those differences, the closest equivalent of not having EXPLICIT_VHOSTS is just to give it the same context creation info without zeroing it down.  You'll probably want to enable TLS in the context and serve wss in a real server, unless it's behind a reverse proxy.

> I'm a little confused as to whether or not I should have a single protocol structure and if info.protocols should point to the same structure for both vhosts. Since there is a single event loop I am assuming that they should be the same.

No they don't have to be the same.  It's better to use info.pprotocols (notice the extra `p`) and give it a NULL-terminated array of pointers to lws_protocols... then you can keep everything private except export the lws_protocols, eg, each in its own source file, everything declared static in there except the lws_protocols for it which you extern into your main.c or whatever.

Lws can handle as many listen sockets / vhosts as you want on the one event loop in the same lws_context, same as handling as many accepted connections on any mixture of the vhosts as you want all on the same loop as the listen sockets live on too.  So you almost always only need one lws_context, one event loop in that and then as many of everything else living on that as you like.  This leads to good behaviours like being able to confidently manage lifecycle of, eg, raw wsi bound to ws wsi without worrying about locking... both are - everything in lws is - serviced sequentially on the same thread and can never conflict.  So they can touch each others' pss members etc cleanly.  (If you do that, you must take care that any real network connection can close and destroy the wsi any time around the event loop, and must then detach from its partner or otherwise deal with the lifecycle of the bound connection... but that can never happen while you are in an lws callback, only after you returned to the event loop.)

**Comment by mfallavol** [comment]
Thanks.  rebuilding with -DLWS_WITH_SECURE_STREAMS=0 did the trick.  I seem to still be able to communicate with the websocket server just fine however when I connected and sent a miscellaneous string to my "raw" port using `nc localhost 5310` I saw the following while handling the LWS_CALLBACK_RAW_RX callback:

1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: .....
Sec-WebSocket-Protocol: ...
uWebSockets: 18`

Is this expected with `info.options=LWS_SERVER_OPTION_ONLY_RAW`?  I thought the reason for the option was to disable attempts to impose websocket protocol.
